### ISSUE TITLE
Create an API for the Source Interface

### ISSUE THREAD

**Issue by micahflee** [freedomofpress/securedrop#91]
I've talked with people from the Guardian Project and WITNESS about building SecureDrop support into [InformaCam](https://guardianproject.info/apps/informacam/).

InformaCam is an app that lets you take photos or videos, attach a bunch of metadata to it, and securely send it to a media organization over Tor.

They could try to hack it together by making the proper GET and POST requests, but I'm sure that would quickly break as we update SecureDrop. In order to make things less brittle, we should build an API for them.

**Comment by garrettr** [comment]
Informacam is still under active development, so we should consider revisiting this.

**Comment by heartsucker** [comment]
Hey y'all. Coming back to this after tonight's discussion, one question. Should the API allow for anonymous users? As in people who want to "fire and forget?" Or should we assume all users want to be able to go to a laptop later and check in with the journalists?

Note: this probably depends on #2057 and the use of `Blueprint`s so we can add custom wrappers and error handlers more uniformly to each of the two halves of the source interface.

# API Proposal

Unless otherwise noted:
- Accepts `application/json`
- Emits `application/json`
- Versioned with `/api/v1/`
- Authorized with headers

## Errors

All 4xx and 5xx errors respond with

```json
{
  "error": "<message>"
}
```

## Endpoints

### Auth / register: `/auth`

#### New Account: `POST /auth/new`

This is the only endpoint the does not require auth.

Response:

```json
{
  "codename": "asylum napkin combat",
  "token": {
    "value": "dK1QOYfEW3QyiKIs0CQHgxOAKMO60ecgJQa/EoQ01P4=",
    "expires": "2017-08-01T12:43:56Z"
  }
}
```

Where `codename` is the codename that can be used to log in to the web interface and `token` is an a bearer token to be set in the headers for future requests.

#### New Token: `POST /auth/token?codename=<codename>&expires=<expires>`

```json
{
  "value": "dK1QOYfEW3QyiKIs0CQHgxOAKMO60ecgJQa/EoQ01P4=",
  "expires": "2017-08-01T12:43:56Z"
}
```

Returns an error if `codename` is not found. Expires is capped at a max and overridden if it exceeds this.

#### Deauthorize token: `DELETE /auth/token/<token>`

#### Deauthorize all tokens: `DELETE /auth/tokens`

### Account: `/account`

#### Delete Account: `DELETE /account`

Response: `204 No Content`

### Submissions: `/submission`

This section might need future-proofing in the even that we want a source to be able to have multiple conversations at once. This may be reasonable if they want to submit different docs to different journalists in an org.

#### Submit new files: `POST /submission/new`

Accepts: `multipart/form-data` with 1 message (optional) and some N files

Response: `204 No Content`

#### Submit a new stream of data: `POST /submission/stream`

TODO: Needs research on how to best handle streaming media (video).

#### List Replies: `GET /submission/replies?offset=<offset>&max_results=<max_results>`

Messages are sorted by most recent. Results are paginated.

```json
{
  "messages": [
    {
      "id": "<message_id>",
      "from": "<journalist_display_name>",
      "message": "<content>",
      "timestamp": "<timestamp>"
    }
  ],
  "prev_page": "<url_path>",
  "next_page": "<url_path>"
}
```

`id` is a UUID (**not** an int). `timestamp` is the rounded timestamp like in the source interface. `from` is an optional value present in responses. `next_page` and `prev_page` are an optional values that point to the next set of messages in the series.

#### Delete a message: `DELETE /submission/reply/<message_id>`

Response: `204 No Content`

#### Delete all messages: `DELETE /submission/replies`

Response: `204 No Content`

**Comment by dachary** [comment]
Note to self: related to https://github.com/freedomofpress/securedrop/issues/1761

**Comment by redshiftzero** [comment]
Excellent, thanks for writing this up @heartsucker - this looks great on first inspection. Some thoughts:

* I think we should allow anonymous users (but keeping what you have above in case that sources do want to carry on a conversation with an organization). I expect a lot of the people that would want to use an API like this to autoupload photos/videos to a particular organization's SecureDrop might not care about conversations with sources (e.g. [this](https://openoversight.lucyparsonslabs.com/submit) [project](https://github.com/lucyparsons/OpenOversight/) encouraging anonymous submissions of photos of police officers doesn't care much about talking with the submitters, they just want the photos).

* I would say that it's fine to ignore the case of a source having multiple conversations for now. Implementing the submission of content to particular journalists (i.e. encryption to multiple keys) or private conversations between sources and particular journalists would be a big change for SecureDrop and I think would justify `/api/v2/`.

* Re `GET /submission/replies?offset=<offset>&max_results=<max_results>`: While we don't have a `journalist_display_name` (for now this is basically just their `username` because we don't store any other journalist information) or show the `username` anywhere to sources, I think this is likely pretty confusing because a source doesn't know who they are talking to, or even if it's one person or multiple people. 

Also, seeing the pagination on the list replies endpoint reminded me that the SecureDrop frontend has no pagination, but we definitely should - made a followup issue for that #2067

**Comment by heartsucker** [comment]
Ok, I'm going to apply the edit above since that will make it easier to reference later.

**Comment by heartsucker** [comment]
I didn't build in support for throttled key generation. I know at somepoint there was discussion about whether or not it made sense to keep using GPG on the source interface. Has that discussion progressed at all?

**Comment by redshiftzero** [comment]
It's safe to assume we are going to keep using GPG for the near future - like many people who use GPG we keep talking about moving to something else but we don't actually have a solid plan of what to move to... :woman_shrugging: 

One other thing: we have this new JSON endpoint in the SecureDrop 4.x series (`/metadata` on the source interface) that provides the fingerprint of the submission key - in the source API we should have a JSON endpoint that has the full public key for download. The reason being if we eventually have submissions coming in through an API like what is proposed in this ticket, we could have submissions get encrypted to the PGP public key client-side (see very long discussion in #92).

**Comment by heartsucker** [comment]
> we don't actually have a solid plan of what to move to

Clearly the solution is to roll our own crypto :)

But yeah, I'll make some changes above.

**Comment by heartsucker** [comment]
Further discussion: Since we'd be serving the GPG key to consumers of the API, would it be worth sniffing the incoming messages to ensure that they are GPG encrypted and then reject those that aren't? I've been playing around with `pyasn1` for something at work and think I could reliably parse the beginning of a message to ensure that it's GPG encrypted. Or we could just do that lazy thing and assume that everything is plaintext and re-encrypt serverside.

**Comment by redshiftzero** [comment]
this would also be useful for allowing people to upload using https://open-archive.org/save/
